core-services/workflow/src/main/java/org/silverpeas/core/workflow/engine/WorkflowTools.java

```diff
@@ -0,0 +1,334 @@
+          if (message == null || message.length() == 0) {
+            message = action.getDescription("", "");
+          }
+
+          // check if sender has been hardcoded in the model
+          String senderId = notifiedUsers.getSenderId();
+          User forcedUser = null;
+          if (senderId != null) {
+            try {
+              forcedUser = userManager.getUser(senderId);
+            } catch (WorkflowException we) {
+              SilverLogger.getLogger(WorkflowTools.class)
+                  .error("Error while processing {0}: impossible to find the sender with id {1}",
+                      event.getActionName(), senderId);
+            }
+          }
+
+          User sender = (forcedUser == null) ? event.getUser() : forcedUser;
+          taskManager.notifyActor(tasks[i], sender, actors[i].getUser(), message);
+        }
+      }
+    } catch (Exception e) {
+      // change the action status of the step : Process Failed
+      step.setActionStatus(-1);
+      instance.updateHistoryStep(step);
+      throw new WorkflowException("WorkflowEngineThread.process",
+          "workflowEngine.EX_ERR_PROCESS_EVENT", e);
+    }
+
+    // change the action status of the step
+    step.setActionStatus(1); // Processed
+    instance.updateHistoryStep(step);
+
+    // Compute states and affectations
+    computeStates(instance, event, eligibleStates, oldActiveStates);
+
+    // change the action status of the step
+    // Affectations done
+    step.setActionStatus(2);
+    instance.updateHistoryStep(step);
+
+    // Process external actions
+    processTriggers(consequence, instance, event);
+
+    return false;
+  }
+
+  private static void processTriggers(Consequence consequence, UpdatableProcessInstance instance,
+      GenericEvent event) {
+    if (consequence != null) {
+      Iterator<Trigger> triggers = consequence.getTriggers().iterateTrigger();
+      while (triggers.hasNext()) {
+        Trigger trigger = triggers.next();
+        if (trigger != null) {
+          try {
+            ExternalAction externalAction = (ExternalAction) Class.forName(trigger.getClassName()).
+                newInstance();
+            externalAction.setProcessInstance(instance);
+            externalAction.setEvent(event);
+            externalAction.setTrigger(trigger);
+            externalAction.execute();
+          } catch (Exception e) {
+            SilverLogger.getLogger(WorkflowTools.class)
+                .error("Error while processing triggers: action = {0}, trigger = {1}",
+                    new String[] {event.getActionName(), trigger.getName()}, e);
+          }
+        }
+      }
+    }
+  }
+
+  /**
+   * Compute the new active states and updates affections
+   * @param event the task done event
+   */
+  private static void computeStates(UpdatableProcessInstance instance, GenericEvent event,
+      Hashtable<String, String> eligibleStates, Hashtable<String, String> oldActiveStates)
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Replace the synchronized class "Hashtable" by an unsynchronized one such as "HashMap". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://integration.silverpeas.org/sonar/coding_rules#rule_key=squid%3AS1149)
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Replace the synchronized class "Hashtable" by an unsynchronized one such as "HashMap". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://integration.silverpeas.org/sonar/coding_rules#rule_key=squid%3AS1149)